File: packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/CozyMarkerInterpolator.m
//
//  CozyMarkerInterpolator.m
//  Pods
//
//  Created by Pietro Domingues on 26/09/23.
//

#import <UIKit/UIKit.h>
#import "CozyMarkerInterpolator.h"
#import "CozyMarkerElements.h"

@implementation CozyMarkerInterpolator

- (CozyMarkerElements *)getInterpolatedMarkerElementsWithStart:(CozyMarkerElements *)start end:(CozyMarkerElements *)end step:(CGFloat)step {
    CozyMarkerElement *interpolatedCanvas = [self interpolateCozyMarkerElementWithStart:start.canvas end:end.canvas step:step];
    CozyMarkerElement *interpolatedBubble = [self interpolateCozyMarkerElementWithStart:start.bubble end:end.bubble step:step];
    CozyMarkerElement *interpolatedIcon = [self interpolateCozyMarkerElementWithStart:start.icon end:end.icon step:step];
    CozyMarkerElement *interpolatedIconCircle = [self interpolateCozyMarkerElementWithStart:start.iconCircle end:end.iconCircle step:step];
    CozyMarkerElement *interpolatedPointer = [self interpolateCozyMarkerElementWithStart:start.pointer end:end.pointer step:step];

    CozyMarkerElement *interpolatedLabel = [self interpolateCozyMarkerElementWithStart:start.labels[0] end:end.labels[0] step:step];
    NSArray<CozyMarkerElement *> *interpolatedLabels = [self interpolateLabelWithInterpolatedLabel:interpolatedLabel startText:start.labels[0].data endText:end.labels[0].data step:step];
    interpolatedIcon.data = [self interpolateIconsWithStart:start.icon.data end:end.icon.data step:step];

    return [[CozyMarkerElements alloc] initWithCanvas:interpolatedCanvas bubble:interpolatedBubble labels:interpolatedLabels icon:interpolatedIcon iconCircle:interpolatedIconCircle pointer:interpolatedPointer];
}

- (CozyMarkerElement *)interpolateCozyMarkerElementWithStart:(CozyMarkerElement *)start end:(CozyMarkerElement *)end step:(CGFloat)step {
    CGRect interpolatedBounds = CGRectMake(
        [self interpolate:start.bounds.origin.x end:end.bounds.origin.x step:step],
        [self interpolate:start.bounds.origin.y end:end.bounds.origin.y step:step],
        [self interpolate:start.bounds.size.width end:end.bounds.size.width step:step],
        [self interpolate:start.bounds.size.height end:end.bounds.size.height step:step]
    );

    UIColor *interpolatedFillColor = [self interpolateColorWithStart:start.fillColor end:end.fillColor step:step];
    UIColor *interpolatedStrokeColor = [self interpolateColorWithStart:start.strokeColor end:end.strokeColor step:step];
    CGFloat interpolatedAlpha = [self interpolate:start.alpha end:end.alpha step:step];

    return [[CozyMarkerElement alloc] initWithBounds:interpolatedBounds fillColor:interpolatedFillColor strokeColor:interpolatedStrokeColor alpha:interpolatedAlpha data:nil];
}

- (CGFloat)interpolate:(CGFloat)start end:(CGFloat)end step:(CGFloat)step {
    return start + (end - start) * step;
}

- (UIColor *)interpolateColorWithStart:(UIColor *)start end:(UIColor *)end step:(CGFloat)step {
    CGFloat startRed, startGreen, startBlue, startAlpha;
    CGFloat endRed, endGreen, endBlue, endAlpha;
    
    [start getRed:&startRed green:&startGreen blue:&startBlue alpha:&startAlpha];
    [end getRed:&endRed green:&endGreen blue:&endBlue alpha:&endAlpha];
    
    CGFloat interpolatedRed = [self interpolate:startRed end:endRed step:step];
    CGFloat interpolatedGreen = [self interpolate:startGreen end:endGreen step:step];
    CGFloat interpolatedBlue = [self interpolate:startBlue end:endBlue step:step];
    CGFloat interpolatedAlpha = [self interpolate:startAlpha end:endAlpha step:step];
    
    return [UIColor colorWithRed:interpolatedRed green:interpolatedGreen blue:interpolatedBlue alpha:interpolatedAlpha];
}

- (NSArray<CozyMarkerElement *> *)interpolateLabelWithInterpolatedLabel:(CozyMarkerElement *)interpolatedLabel startText:(NSString *)startText endText:(NSString *)endText step:(CGFloat)step {
    if ([startText isEqualToString:endText]) {
        return @[
            [[CozyMarkerElement alloc] initWithBounds:interpolatedLabel.bounds fillColor:interpolatedLabel.fillColor strokeColor:interpolatedLabel.strokeColor alpha:interpolatedLabel.alpha data:startText]
        ];
    }
    
    CGFloat startTextAlpha = 1.0 - step;
    CGFloat endTextAlpha = step;
    
    CozyMarkerElement *startLabelElement = [[CozyMarkerElement alloc] initWithBounds:interpolatedLabel.bounds fillColor:interpolatedLabel.fillColor strokeColor:interpolatedLabel.strokeColor alpha:startTextAlpha data:startText];
    CozyMarkerElement *endLabelElement = [[CozyMarkerElement alloc] initWithBounds:interpolatedLabel.bounds fillColor:interpolatedLabel.fillColor strokeColor:interpolatedLabel.strokeColor alpha:endTextAlpha data:endText];
    
    return @[startLabelElement, endLabelElement];
}

- (NSString *)interpolateIconsWithStart:(NSString *)startIcon end:(NSString *)endIcon step:(CGFloat)step {
    if (startIcon) {
        return startIcon;
    }
    if (endIcon) {
        return endIcon;
    }
    return nil;
}
@end

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/cozy/CozyMarkerInterpolator.java
package io.flutter.plugins.googlemaps.cozy;

import android.graphics.Bitmap;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Typeface;
import android.graphics.RectF;

class CozyMarkerInterpolator {
    private final Typeface font;

    CozyMarkerInterpolator(Typeface font) {
        this.font = font;
    }

    public CozyMarkerElements getInterpolatedMarkerObjects(CozyMarkerElements startCozyMarkerElements, CozyMarkerElements endCozyMarkerElements, float step) {
        // interpolate geometry
        CozyMarkerElement interpolatedCanvas = interpolateCozyMarkerElement(startCozyMarkerElements.canvas, endCozyMarkerElements.canvas, step);
        CozyMarkerElement interpolatedBubble = interpolateCozyMarkerElement(startCozyMarkerElements.bubble, endCozyMarkerElements.bubble, step);
        CozyMarkerElement interpolatedIcon = interpolateCozyMarkerElement(startCozyMarkerElements.icons[0], endCozyMarkerElements.icons[0], step);
        CozyMarkerElement interpolatedIconCircle = interpolateCozyMarkerElement(startCozyMarkerElements.iconCircle, endCozyMarkerElements.iconCircle, step);
        CozyMarkerElement interpolatedPointer = interpolateCozyMarkerElement(startCozyMarkerElements.pointer, endCozyMarkerElements.pointer, step);
        CozyMarkerElement interpolatedCounterBubble = interpolateCozyMarkerElement(startCozyMarkerElements.counterBubble, endCozyMarkerElements.counterBubble, step);

        CozyMarkerElement interpolatedLabel = interpolateCozyMarkerElement(startCozyMarkerElements.labels[0], endCozyMarkerElements.labels[0], step);

        CozyMarkerElement[] interpolatedBubbles = interpolateLabel(interpolatedBubble, (String) startCozyMarkerElements.bubble.data, (String) endCozyMarkerElements.bubble.data, step);
        CozyMarkerElement[] interpolatedLabels = interpolateLabel(interpolatedLabel, (String) startCozyMarkerElements.labels[0].data, (String) endCozyMarkerElements.labels[0].data, step);
        CozyMarkerElement[] interpolatedIcons = interpolateIcons(interpolatedIcon, (Bitmap) startCozyMarkerElements.icons[0].data, (Bitmap) endCozyMarkerElements.icons[0].data, step);

        return new CozyMarkerElements(interpolatedCanvas, interpolatedBubble, interpolatedCounterBubble, interpolatedLabels, interpolatedIcons, interpolatedIconCircle, interpolatedPointer);
    }

    private CozyMarkerElement[] interpolateLabel(CozyMarkerElement interpolatedLabel, String startText, String endText, float step) {
        if (startText == null || endText == null || startText.equals(endText)) {
            return new CozyMarkerElement[]{
                    new CozyMarkerElement(
                            interpolatedLabel.bounds,
                            interpolatedLabel.fillColor,
                            interpolatedLabel.strokeColor,
                            interpolatedLabel.alpha,
                            startText
                    )
            };
        }

        // Calculate the size of string
        Paint textPaint = new Paint();
        textPaint.setTypeface(font);
        textPaint.setTextSize(interpolatedLabel.bounds.height());
        textPaint.setAntiAlias(true);
        textPaint.setTextAlign(Paint.Align.LEFT);

        int startTextCharactersNumber = textPaint.breakText(startText, true, interpolatedLabel.bounds.width() * 1.2f, null);
        int endTextCharactersNumber = textPaint.breakText(endText, true, interpolatedLabel.bounds.width() * 1.2f, null);

        String clippedStartText = startText.substring(0, startTextCharactersNumber);
        String clippedEndText = endText.substring(0, endTextCharactersNumber);

        float startTextAlpha = 1.0f - step;
        float endTextAlpha = step;

        return new CozyMarkerElement[]{
                new CozyMarkerElement(
                        interpolatedLabel.bounds,
                        interpolatedLabel.fillColor,
                        interpolatedLabel.strokeColor,
                        startTextAlpha,
                        clippedStartText
                ),
                new CozyMarkerElement(
                        interpolatedLabel.bounds,
                        interpolatedLabel.fillColor,
                        interpolatedLabel.strokeColor,
                        endTextAlpha,
                        clippedEndText
                )
        };
    }

    private CozyMarkerElement[] interpolateIcons(CozyMarkerElement interpolatedIcon, Bitmap startIcon, Bitmap endIcon, float step) {
        if (startIcon == null && endIcon == null) {
            return new CozyMarkerElement[]{interpolatedIcon};
        }
        if (startIcon == null) {
            return new CozyMarkerElement[]{
                    new CozyMarkerElement(
                            interpolatedIcon.bounds,
                            interpolatedIcon.fillColor,
                            interpolatedIcon.strokeColor,
                            interpolatedIcon.alpha,
                            endIcon
                    )
            };
        }
        if (endIcon == null) {
            return new CozyMarkerElement[]{
                    new CozyMarkerElement(
                            interpolatedIcon.bounds,
                            interpolatedIcon.fillColor,
                            interpolatedIcon.strokeColor,
                            interpolatedIcon.alpha,
                            startIcon
                    )
            };
        }
        if (startIcon.sameAs(endIcon)) {
            return new CozyMarkerElement[]{
                    new CozyMarkerElement(
                            interpolatedIcon.bounds,
                            interpolatedIcon.fillColor,
                            interpolatedIcon.strokeColor,
                            interpolatedIcon.alpha,
                            startIcon
                    )
            };
        }
        return new CozyMarkerElement[]{
                new CozyMarkerElement(
                        interpolatedIcon.bounds,
                        interpolatedIcon.fillColor,
                        interpolatedIcon.strokeColor,
                        interpolatedIcon.alpha * (1.0f - step),
                        startIcon
                ),
                new CozyMarkerElement(
                        interpolatedIcon.bounds,
                        interpolatedIcon.fillColor,
                        interpolatedIcon.strokeColor,
                        interpolatedIcon.alpha * step,
                        endIcon
                )
        };
    }

    private CozyMarkerElement interpolateCozyMarkerElement(CozyMarkerElement startCozyMarkerElement, CozyMarkerElement endCozyMarkerElement, float step) {
        if (startCozyMarkerElement == null || endCozyMarkerElement == null || startCozyMarkerElement.bounds == null || endCozyMarkerElement.bounds == null) {
            return null;
        }

        RectF interpolatedBounds = new RectF();
        interpolatedBounds.left = interpolate(startCozyMarkerElement.bounds.left, endCozyMarkerElement.bounds.left, step);
        interpolatedBounds.top = interpolate(startCozyMarkerElement.bounds.top, endCozyMarkerElement.bounds.top, step);
        interpolatedBounds.right = interpolate(startCozyMarkerElement.bounds.right, endCozyMarkerElement.bounds.right, step);
        interpolatedBounds.bottom = interpolate(startCozyMarkerElement.bounds.bottom, endCozyMarkerElement.bounds.bottom, step);

        int interpolatedFillColor = interpolateColor(startCozyMarkerElement.fillColor, endCozyMarkerElement.fillColor, step);
        int interpolatedStrokeColor = interpolateColor(startCozyMarkerElement.strokeColor, endCozyMarkerElement.strokeColor, step);
        float interpolatedAlpha = interpolate(startCozyMarkerElement.alpha, endCozyMarkerElement.alpha, step);

        return new CozyMarkerElement(interpolatedBounds, interpolatedFillColor, interpolatedStrokeColor, interpolatedAlpha);
    }

    private float interpolate(float start, float end, float step) {
        return start + (end - start) * step;
    }

    private int interpolateColor(int start, int end, float step) {
        int startRed = (start >> 16) & 0xff;
        int startGreen = (start >> 8) & 0xff;
        int startBlue = start & 0xff;

        int endRed = (end >> 16) & 0xff;
        int endGreen = (end >> 8) & 0xff;
        int endBlue = end & 0xff;

        int interpolatedRed = (int) interpolate(startRed, endRed, step);
        int interpolatedGreen = (int) interpolate(startGreen, endGreen, step);
        int interpolatedBlue = (int) interpolate(startBlue, endBlue, step);

        return Color.rgb(interpolatedRed, interpolatedGreen, interpolatedBlue);
    }
}